metricbeat/module/elasticsearch/cluster_stats/data_xpack.go

```diff
@@ -0,0 +1,238 @@
+					"percent": c.Int("percent"),
+				}),
+				"open_file_descriptors": c.Dict("open_file_descriptors", s.Schema{
+					"min": c.Int("min"),
+					"max": c.Int("max"),
+					"avg": c.Int("avg"),
+				}),
+			}),
+			"jvm": c.Dict("jvm", s.Schema{
+				"max_uptime_in_millis": c.Int("max_uptime_in_millis"),
+				"mem": c.Dict("mem", s.Schema{
+					"heap_used_in_bytes": c.Int("heap_used_in_bytes"),
+					"heap_max_in_bytes":  c.Int("heap_max_in_bytes"),
+				}),
+				"threads": c.Int("threads"),
+			}),
+			"fs": c.Dict("fs", s.Schema{
+				"total_in_bytes":     c.Int("total_in_bytes"),
+				"free_in_bytes":      c.Int("free_in_bytes"),
+				"available_in_bytes": c.Int("available_in_bytes"),
+			}),
+		}),
+	}
+)
+
+func passthruField(fieldPath string, sourceData, targetData common.MapStr) error {
+	fieldValue, err := sourceData.GetValue(fieldPath)
+	if err != nil {
+		return elastic.MakeErrorForMissingField(fieldPath, elastic.Elasticsearch)
+	}
+
+	targetData.Put(fieldPath, fieldValue)
+	return nil
+}
+
+func eventMappingXPack(r mb.ReporterV2, m *MetricSet, content []byte) error {
+	var data map[string]interface{}
+	err := json.Unmarshal(content, &data)
+	if err != nil {
+		return err
+	}
+
+	clusterStats, err := clusterStatsSchema.Apply(data)
+	if err != nil {
+		return err
+	}
+
+	dataMS := common.MapStr(data)
+
+	passthruFields := []string{
+		"indices.segments.file_sizes", // object with dynamic keys
+		"nodes.versions", // array of strings
+		"nodes.os.names", // array of objects
+		"nodes.jvm.versions", // array of objects
+		"nodes.plugins", // array of objects
+		"nodes.network_types", // object with dynamic keys
+	}
+	for _, fieldPath := range passthruFields {
+		if err = passthruField(fieldPath, dataMS, clusterStats); err != nil {
+			return err
+		}
+	}
+
+	clusterName, ok := data["cluster_name"].(string)
+	if !ok {
+		return elastic.MakeErrorForMissingField("cluster_name", elastic.Elasticsearch)
+	}
+
+	info, err := elasticsearch.GetInfo(m.HTTP, m.HTTP.GetURI())
+	if err != nil {
+		return err
+	}
+
+	license, err := elasticsearch.GetLicense(m.HTTP, m.HTTP.GetURI())
+	if err != nil {
+		return err
+	}
+
+	// TODO: Inject `cluster_needs_tls` field under license object
+
+	clusterState, err := elasticsearch.GetClusterState(m.HTTP, m.HTTP.GetURI())
+	if err != nil {
+		return err
+	}
+
+	if err = passthruField("status", dataMS, clusterState); err != nil {
+		return err
+	}
+
+	// TODO: Compute and inject `node_hash` field under clusterState object
+
+	stackStats, err := elasticsearch.GetStackStats(m.HTTP, m.HTTP.GetURI())
+	if err != nil {
+		return err
+	}
+
+	// TODO: Inject `apm.found` field under stackStats object
+
+	event := mb.Event{}
+	event.RootFields = common.MapStr{
+		"cluster_uuid":  info.ClusterID,
+		"cluster_name":  clusterName,
+		"timestamp":     common.Time(time.Now()),
+		"interval_ms":   m.Module().Config().Period / time.Millisecond,
+		"type":          "cluster_stats",
+		"license":       license,
```

Comment:
Is the full license response reported here?

Reply:
Yes, it is the response from the `GET _xpack/license` Elasticsearch API, which looks like this:

```js
{
  "license": {
    "status": "active",
    "uid": "f80c1fb5-75e9-4536-be2d-a768b02abb46",
    "type": "basic",
    "issue_date": "2018-08-06T23:47:09.619Z",
    "issue_date_in_millis": 1533599229619,
    "max_nodes": 1000,
    "issued_to": "elasticsearch",
    "issuer": "elasticsearch",
    "start_date_in_millis": -1
  }
}
```